diff -NurpX linux-4.9.24-pax/Documentation/dontdiff linux-4.9.24/arch/parisc/kernel/sys_parisc.c linux-4.9.24-pax/arch/parisc/kernel/sys_parisc.c
--- linux-4.9.24/arch/parisc/kernel/sys_parisc.c	2016-05-22 01:55:28.571569665 +0200
+++ linux-4.9.24-pax/arch/parisc/kernel/sys_parisc.c	2017-01-01 22:57:10.569520647 +0100
@@ -109,6 +109,10 @@ unsigned long arch_get_unmapped_area(str
 		goto found_addr;
 	}
 
+#ifdef CONFIG_PAX_RANDMMAP
+	if (!(mm->pax_flags & MF_PAX_RANDMMAP))
+#endif
+
 	if (addr) {
 		if (do_color_align && last_mmap)
 			addr = COLOR_ALIGN(addr, last_mmap, pgoff);
@@ -170,6 +174,10 @@ arch_get_unmapped_area_topdown(struct fi
 	}
 
 	/* requesting a specific address */
+#ifdef CONFIG_PAX_RANDMMAP
+	if (!(mm->pax_flags & MF_PAX_RANDMMAP))
+#endif
+
 	if (addr) {
 		if (do_color_align && last_mmap)
 			addr = COLOR_ALIGN(addr, last_mmap, pgoff);
@@ -252,6 +260,13 @@ void arch_pick_mmap_layout(struct mm_str
 	mm->mmap_legacy_base = mmap_legacy_base();
 	mm->mmap_base = mmap_upper_limit();
 
+#ifdef CONFIG_PAX_RANDMMAP
+	if (mm->pax_flags & MF_PAX_RANDMMAP) {
+		mm->mmap_legacy_base += mm->delta_mmap;
+		mm->mmap_base -= mm->delta_mmap + mm->delta_stack;
+	}
+#endif
+
 	if (mmap_is_legacy()) {
 		mm->mmap_base = mm->mmap_legacy_base;
 		mm->get_unmapped_area = arch_get_unmapped_area;
