diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index cf57a77..ab33bd2 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -906,6 +906,7 @@ config RELOCATABLE
 
 config RANDOMIZE_BASE
 	bool "Randomize the address of the kernel image"
+	depends on BROKEN_SECURITY
 	select ARM64_MODULE_PLTS if MODULES
 	select RELOCATABLE
 	help
diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index 5a4f2eb..93c4de6 100644
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -2596,7 +2597,7 @@ config RELOCATION_TABLE_SIZE
 
 config RANDOMIZE_BASE
 	bool "Randomize the address of the kernel image"
-	depends on RELOCATABLE
+	depends on RELOCATABLE && BROKEN_SECURITY
 	---help---
 	   Randomizes the physical and virtual address at which the
 	   kernel image is loaded, as a security feature that
diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index bada636..1775eac 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -1934,7 +1940,7 @@ config RELOCATABLE
 
 config RANDOMIZE_BASE
 	bool "Randomize the address of the kernel image (KASLR)"
-	depends on RELOCATABLE
+	depends on RELOCATABLE && BROKEN_SECURITY
 	default n
 	---help---
 	  In support of Kernel Address Space Layout Randomization (KASLR),
