diff -NurpX linux-4.9.24-pax/Documentation/dontdiff linux-4.9.24/arch/arm/kernel/process.c linux-4.9.24-pax/arch/arm/kernel/process.c
--- linux-4.9.24/arch/arm/kernel/process.c	2016-12-13 12:08:18.122317736 +0100
+++ linux-4.9.24-pax/arch/arm/kernel/process.c	2017-01-01 22:57:10.533520559 +0100
@@ -336,7 +336,7 @@ static struct vm_area_struct gate_vma =
 
 static int __init gate_vma_init(void)
 {
-	gate_vma.vm_page_prot = PAGE_READONLY_EXEC;
+	gate_vma.vm_page_prot = vm_get_page_prot(gate_vma.vm_flags);
 	return 0;
 }
 arch_initcall(gate_vma_init);
diff -NurpX linux-4.9.24-pax/Documentation/dontdiff linux-4.9.24/arch/ia64/mm/init.c linux-4.9.24-pax/arch/ia64/mm/init.c
--- linux-4.9.24/arch/ia64/mm/init.c	2015-11-03 01:48:49.911328199 +0100
+++ linux-4.9.24-pax/arch/ia64/mm/init.c	2017-01-01 22:57:10.557520617 +0100
@@ -279,7 +292,7 @@ static int __init gate_vma_init(void)
 	gate_vma.vm_start = FIXADDR_USER_START;
 	gate_vma.vm_end = FIXADDR_USER_END;
 	gate_vma.vm_flags = VM_READ | VM_MAYREAD | VM_EXEC | VM_MAYEXEC;
-	gate_vma.vm_page_prot = __P101;
+	gate_vma.vm_page_prot = vm_get_page_prot(gate_vma.vm_flags);
 
 	return 0;
 }
