diff -NurpX linux-4.9.24-pax/Documentation/dontdiff linux-4.9.24/arch/powerpc/kernel/Makefile linux-4.9.24-pax/arch/powerpc/kernel/Makefile
--- linux-4.9.24/arch/powerpc/kernel/Makefile	2016-12-13 12:08:23.902357033 +0100
+++ linux-4.9.24-pax/arch/powerpc/kernel/Makefile	2017-01-01 22:57:10.577520667 +0100
@@ -15,7 +15,7 @@ CFLAGS_btext.o		+= -fPIC
 endif
 
 CFLAGS_cputable.o += $(DISABLE_LATENT_ENTROPY_PLUGIN)
-CFLAGS_init.o += $(DISABLE_LATENT_ENTROPY_PLUGIN)
+CFLAGS_prom_init.o += $(DISABLE_LATENT_ENTROPY_PLUGIN)
 CFLAGS_btext.o += $(DISABLE_LATENT_ENTROPY_PLUGIN)
 CFLAGS_prom.o += $(DISABLE_LATENT_ENTROPY_PLUGIN)
 
