diff -NurpX linux-4.9.24-pax/Documentation/dontdiff linux-4.9.24/arch/x86/kernel/dumpstack_32.c linux-4.9.24-pax/arch/x86/kernel/dumpstack_32.c
--- linux-4.9.24/arch/x86/kernel/dumpstack_32.c	2016-12-13 12:08:27.286380068 +0100
+++ linux-4.9.24-pax/arch/x86/kernel/dumpstack_32.c	2017-01-01 22:57:10.653520853 +0100
@@ -205,3 +207,15 @@ int is_valid_bugaddr(unsigned long ip)
 
 	return ud2 == 0x0b0f;
 }
+
+#ifdef CONFIG_PAX_MEMORY_STACKLEAK
+void __used pax_check_alloca(unsigned long size)
+{
+	unsigned long sp = (unsigned long)&sp, stack_left;
+
+	/* all kernel stacks are of the same size */
+	stack_left = sp & (THREAD_SIZE - 1);
+	BUG_ON(stack_left < 256 || size >= stack_left - 256);
+}
+EXPORT_SYMBOL(pax_check_alloca);
+#endif
diff -NurpX linux-4.9.24-pax/Documentation/dontdiff linux-4.9.24/arch/x86/kernel/dumpstack_64.c linux-4.9.24-pax/arch/x86/kernel/dumpstack_64.c
--- linux-4.9.24/arch/x86/kernel/dumpstack_64.c	2016-12-13 12:08:27.290380095 +0100
+++ linux-4.9.24-pax/arch/x86/kernel/dumpstack_64.c	2017-02-27 18:31:38.864954256 +0100
diff -NurpX linux-4.9.24-pax/Documentation/dontdiff linux-4.9.24/arch/x86/kernel/dumpstack_64.c linux-4.9.24-pax/arch/x86/kernel/dumpstack_64.c
--- linux-4.9.24/arch/x86/kernel/dumpstack_64.c	2016-12-13 12:08:27.290380095 +0100
+++ linux-4.9.24-pax/arch/x86/kernel/dumpstack_64.c	2017-02-27 18:31:38.864954256 +0100
@@ -242,3 +242,36 @@ int is_valid_bugaddr(unsigned long ip)
 
 	return ud2 == 0x0b0f;
 }
+
+#ifdef CONFIG_PAX_MEMORY_STACKLEAK
+void __used pax_check_alloca(unsigned long size)
+{
+	struct stack_info stack_info = {0};
+	unsigned long visit_mask = 0;
+	unsigned long sp = (unsigned long)&sp;
+	unsigned long stack_left;
+
+	BUG_ON(get_stack_info(&sp, current, &stack_info, &visit_mask));
+
+	switch (stack_info.type) {
+	case STACK_TYPE_TASK:
+		stack_left = sp & (THREAD_SIZE - 1);
+		break;
+
+	case STACK_TYPE_IRQ:
+		stack_left = sp & (IRQ_STACK_SIZE - 1);
+		break;
+
+	case STACK_TYPE_EXCEPTION ... STACK_TYPE_EXCEPTION_LAST:
+		stack_left = sp & (EXCEPTION_STKSZ - 1);
+		break;
+
+	case STACK_TYPE_SOFTIRQ:
+	default:
+		BUG();
+	}
+
+	BUG_ON(stack_left < 256 || size >= stack_left - 256);
+}
+EXPORT_SYMBOL(pax_check_alloca);
+#endif
