diff -NurpX linux-4.9.24-pax/Documentation/dontdiff linux-4.9.24/arch/parisc/mm/fault.c linux-4.9.24-pax/arch/parisc/mm/fault.c
--- linux-4.9.24/arch/parisc/mm/fault.c	2017-04-19 22:08:36.398272579 +0200
+++ linux-4.9.24-pax/arch/parisc/mm/fault.c	2017-04-19 22:11:31.402403274 +0200
@@ -50,7 +51,7 @@ int show_unhandled_signals = 1;
 static unsigned long
 parisc_acctyp(unsigned long code, unsigned int inst)
 {
-	if (code == 6 || code == 16)
+	if (code == 6 || code == 7 || code == 16)
 	    return VM_EXEC;
 
 	switch (inst & 0xf0000000) {
